src/engine.ml

```diff
@@ -1815,26 +1779,70 @@ let incoming ?(is_not_taken = fun _ip -> false) state buf =
                     (* XXX maybe just ignore? *)
                     if Eqaf_cstruct.equal computed_hmac cleartext.hmac then
                       Ok ()
-                    else Error (`Bad_mac (state, computed_hmac, (key, p)))
+                    else
+                      Error
+                        (`Bad_mac
+                          (state, computed_hmac, ((key, p) :> Packet.t)))
                   in
                   match expected_packet state.session ch.transport p with
                   | Error e ->
                       (* XXX: only in udp mode? *)
                       Log.warn (fun m -> m "ignoring bad packet %a" pp_error e);
-                      Ok (state, out, None)
-                  | Ok (session, transport) ->
+                      Ok (state, out, acts)
+                  | Ok (session, transport) -> (
                       let state = { state with session }
                       and ch = { ch with transport } in
-                      Ok (set_ch state ch, out, act))
-              | Packet.Hard_reset_client_v3, Client_tls_crypt _ ->
-                  Error (`No_transition (ch, op, payload))
-              | _control_op, Client_tls_crypt { tls_crypt = tls_crypt, wkc; _ }
-                -> (
-                  let* cleartext, off =
+                      match p with
+                      | `Ack _ -> Ok (set_ch state ch, out, acts)
+                      | `Control (_, (_, _, data)) -> (
```

Comment:
would be nice to sync the tls_auth and tls_crypt* cases:
- verify_incoming data
- process packet
- sign/encrypt outgoing

this way, the logic (such as ``match out' with [] -> [ (`Ack, Cstruct.empty) ] | xs -> xs``) wouldn't need to be duplicated

bonus points if we merge the tls_crypt_v2 with the `wkc` into the same shape